src/QsCompiler/Transformations/Utilities.cs

```diff
@@ -55,6 +55,30 @@ public static class Utilities
                 QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>.Null
             );
 
+        // All of the resolved types of the given expression must match.
+        public static TypedExpression CreateValueArray(params TypedExpression[] expressions)
+        {
+            ResolvedType type = null;
+            if (expressions.Any())
+            {
+                type = expressions.First().ResolvedType;
+                QsCompilerError.Verify(expressions.All(expr => expr.ResolvedType.Equals(type)));
```

Comment:
This will fail in general.

Reply:
Isn't ResolvedType compared with structural comparison? Doesn't that mean that it is generally safe to compare two ResolvedType objects with the .Equals method?

Reply:
To initialize and array literal it is sufficient for items to have a common base type. Since our arrays are immutable they can be covariant in principle.

Reply:
This is being track in a separate task.